src/handlers/generate-erc20-permit.ts

```diff
@@ -1,58 +1,46 @@
 import { PERMIT2_ADDRESS, PermitTransferFrom, SignatureTransfer } from "@uniswap/permit2-sdk";
 import { ethers, keccak256, MaxInt256, parseUnits, toUtf8Bytes } from "ethers";
 import { Context, Logger } from "../types/context";
-import { Permit } from "../types/permits";
+import { Permit, TokenType } from "../types";
 import { decryptKeys } from "../utils/keys";
 import { getPayoutConfigByNetworkId } from "../utils/payoutConfigByNetworkId";
 
-export async function generateErc20PermitSignature(
-  username: string,
-  amount: number,
-  evmNetworkId: number,
-  evmPrivateEncrypted: string,
-  userId: number,
-  walletAddress: string,
-  issueId: number,
-  logger: Logger
-): Promise<Permit>;
-export async function generateErc20PermitSignature(username: string, amount: number, context: Context): Promise<Permit>;
-export async function generateErc20PermitSignature(
-  username: string,
-  amount: number,
-  contextOrNetworkId: Context | number,
-  evmPrivateEncrypted?: string,
-  userId?: number,
-  walletAddress?: string,
-  issueId?: number,
-  logger?: Logger
-): Promise<Permit> {
+export interface Payload {
+  evmNetworkId: number;
+  evmPrivateEncrypted: string;
+  walletAddress: string;
+  issueId: number;
+  logger: Logger;
+}
+
+export async function generateErc20PermitSignature(payload: Payload, userId: number, amount: number): Promise<Permit>;
+export async function generateErc20PermitSignature(context: Context, userId: number, amount: number): Promise<Permit>;
+export async function generateErc20PermitSignature(contextOrPayload: Context | Payload, userId: number, amount: number): Promise<Permit> {
```

Comment:
maybe it'd better to use single object? I don't know why but trying to fit two completely different objects into one variable seems strange to me
```js
{
  userId: number;
  amount: number;
  evmNetworkId: number;
  evmPrivateEncrypted: string;
  walletAddress: string;
  issueId: number;
  logger: Logger;
} & {
  context: Context;
  userId: number;
  amount: number;
}
```

Reply:
It is strange but inherent to TypeScript way of handling function overloads, as it is not available in JavaScript unlike traditional OOP languages. Problem with using `&` is that I would have to honor both object members, and by using `|` I would have errors like `A is not assignable to B` without a cast which defeats the type checks.
By doing it this way I get full proper auto complete and type check, and is displayed as such in modern IDEs
<img width="455" alt="image" src="https://github.com/ubiquibot/permit-generation/assets/9807008/4ca66e7e-9daa-426f-84f7-993b6814a788">